presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java

```diff
@@ -0,0 +1,762 @@
+        }
+
+        int z0 = 0;
+        int z1 = 0;
+        int z2 = 0;
+        int z3 = 0;
+        int z4 = 0;
+        int z5 = 0;
+        int z6 = 0;
+        int z7 = 0;
+
+        if (l0 != 0) {
+            long accumulator = 0;
+            accumulator = (r0 & LONG_MASK) * (l0 & LONG_MASK) + (z0 & LONG_MASK) + (accumulator >>> 32);
+            z0 = (int) accumulator;
+            accumulator = (r1 & LONG_MASK) * (l0 & LONG_MASK) + (z1 & LONG_MASK) + (accumulator >>> 32);
+            z1 = (int) accumulator;
+            accumulator = (r2 & LONG_MASK) * (l0 & LONG_MASK) + (z2 & LONG_MASK) + (accumulator >>> 32);
+            z2 = (int) accumulator;
+            accumulator = (r3 & LONG_MASK) * (l0 & LONG_MASK) + (z3 & LONG_MASK) + (accumulator >>> 32);
+            z3 = (int) accumulator;
+            z4 = (int) (accumulator >>> 32);
+        }
+
+        if (l1 != 0) {
+            long accumulator = 0;
+            accumulator = (r0 & LONG_MASK) * (l1 & LONG_MASK) + (z1 & LONG_MASK) + (accumulator >>> 32);
+            z1 = (int) accumulator;
+            accumulator = (r1 & LONG_MASK) * (l1 & LONG_MASK) + (z2 & LONG_MASK) + (accumulator >>> 32);
+            z2 = (int) accumulator;
+            accumulator = (r2 & LONG_MASK) * (l1 & LONG_MASK) + (z3 & LONG_MASK) + (accumulator >>> 32);
+            z3 = (int) accumulator;
+            accumulator = (r3 & LONG_MASK) * (l1 & LONG_MASK) + (z4 & LONG_MASK) + (accumulator >>> 32);
+            z4 = (int) accumulator;
+            z5 = (int) (accumulator >>> 32);
+        }
+
+        if (l2 != 0) {
+            long accumulator = 0;
+            accumulator = (r0 & LONG_MASK) * (l2 & LONG_MASK) + (z2 & LONG_MASK) + (accumulator >>> 32);
+            z2 = (int) accumulator;
+            accumulator = (r1 & LONG_MASK) * (l2 & LONG_MASK) + (z3 & LONG_MASK) + (accumulator >>> 32);
+            z3 = (int) accumulator;
+            accumulator = (r2 & LONG_MASK) * (l2 & LONG_MASK) + (z4 & LONG_MASK) + (accumulator >>> 32);
+            z4 = (int) accumulator;
+            accumulator = (r3 & LONG_MASK) * (l2 & LONG_MASK) + (z5 & LONG_MASK) + (accumulator >>> 32);
+            z5 = (int) accumulator;
+            z6 = (int) (accumulator >>> 32);
+        }
+
+        if (l3 != 0) {
+            long accumulator = 0;
+            accumulator = (r0 & LONG_MASK) * (l3 & LONG_MASK) + (z3 & LONG_MASK) + (accumulator >>> 32);
+            z3 = (int) accumulator;
+            accumulator = (r1 & LONG_MASK) * (l3 & LONG_MASK) + (z4 & LONG_MASK) + (accumulator >>> 32);
+            z4 = (int) accumulator;
+            accumulator = (r2 & LONG_MASK) * (l3 & LONG_MASK) + (z5 & LONG_MASK) + (accumulator >>> 32);
+            z5 = (int) accumulator;
+            accumulator = (r3 & LONG_MASK) * (l3 & LONG_MASK) + (z6 & LONG_MASK) + (accumulator >>> 32);
+            z6 = (int) accumulator;
+            z7 = (int) (accumulator >>> 32);
+        }
+
+        if (z7 == 0 && z6 == 0 && z5 == 0 && z4 == 0) {
```

Comment:
Since z4, z5, z6, z7 must be 0, it may be possible to get rid of many of the computations above. For instance. there's no reason to split the last accumulator into z6 and z7. Similarly, `(r3 & LONG_MASK) * (l3 & LONG_MASK) + (z6 & LONG_MASK) + (accumulator >>> 32)` is unnecessary, since we could just check whether r3 or l3 are non-zero to bail out, and so on.